### DISCUSSION TITLE
Getting PointerToNowhere when migrating from RefResolver to referencing

### DISCUSSION THREAD

**Discussion by stefanofusai** [python-jsonschema/jsonschema#1313] [Q&A]
I have been trying my best to migrate from RefResolver to the newer referencing lib but with no avail, and I even requested for help in the openapi-schema-validator repository: https://github.com/python-openapi/openapi-schema-validator/issues/203 as I'm using jsonschema together with the openapi-schema-validator library.

I will report the issue I opened there, hoping that someone from your team can better assist me. Thank you for the help.

---

Dear team,
I've been trying my best to migrate my schema validation from RefResolver to the newer referencing library, but with no avail.
I would really appreciate some help or tips to point me in the right direction.

What is working for me, with RefResolver, is the following code:
```py
def validate_schema(
    instance: dict[str, Any],
    schema: dict[str, Any],
    *,
    path: str,
    method: str,
    response: str,
    content: str = "application/json",
) -> None:
    try:
        _methods = schema["paths"][path]

    except KeyError:  # pragma: no cover (should never happen)
        msg = f"Path {path} not found in schema"
        raise ValueError(msg) from None

    try:
        _responses = _methods[method]

    except KeyError:  # pragma: no cover (should never happen)
        msg = f"Method {method} not found in schema"
        raise ValueError(msg) from None

    try:
        _content = _responses["responses"][response]["content"]

    except KeyError:  # pragma: no cover (should never happen)
        msg = f"Response {response} not found in schema"
        raise ValueError(msg) from None

    try:
        _schema = _content[content]["schema"]

    except KeyError:  # pragma: no cover (should never happen)
        msg = f"Content {content} not found in schema"
        raise ValueError(msg) from None

    openapi_schema_validator.validate(
        instance,
        _schema,
        cls=OAS30Validator,
        resolver=RefResolver.from_schema(schema),  # TODO: migrate to referencing lib
    )
```
I then tried to migrate to the referencing lib as mentioned [here](https://openapi-schema-validator.readthedocs.io/en/latest/references.html):
```py
resource = Resource.from_contents(schema, default_specification=DRAFT202012)
registry = Registry().with_resource("", resource)
openapi_schema_validator.validate(instance, _schema, registry=registry)
```
But with no success, as all I get are errors like:
`jsonschema.exceptions._WrappedReferencingError: PointerToNowhere: '/components/schemas/Foo' does not exist within {'$ref': '#/components/schemas/Foo'}`

What am I missing? Thank you for the help.

**Comment by Julian** [comment]
I will try to find some time to look there, but `{'$ref': '#/components/schemas/Foo'}` is never correct. That pointer means "in this document is a key called `components`, and in it is a subkey called `schemas`, and.. etc.

But that's clearly wrong, as that document has one and only one key in it, it's `$ref`.

**Reply by stefanofusai** [reply]
Thank you for the quick reply. FYI i'm using the auto-generated schema from [drf-spectacular](https://github.com/tfranzel/drf-spectacular).
I would be happy to privately share the full schema, or part of it, if you think that can help me in the right direction.

**Reply by Julian** [reply]
(Usually what you need to do is make your `$ref` properly start with whatever URI you are trying to refer to -- e.g. if your root document has `$id: "urn:example:myCoolSchema"` then you will instead want to do `$ref: "urn:example:myCoolSchema#/components/schemas/Foo"` -- but I'll also note that OpenAPI prior to 3.1 had entirely different referencing semantics from JSON Schema, so hopefully you're on the latest version.)

**Reply by stefanofusai** [reply]
I am in fact using OpenAPI 3.1!
```py
openapi_schema_validator.validate(
    instance,
    _schema,
    cls=OAS31Validator,
    resolver=RefResolver.from_schema(schema),  # TODO: migrate to referencing lib
)
```

If it can help you, in this code snippet, here's an example of what the variables eval to during a test run:
- `instance`: `{'pagination': ..., 'results': ...}` # this is the response data
- `_schema`: `{'$ref': '#/components/schemas/MySchemaName'}`
- `schema`: 
```json
{
    "components": {
        "schemas": {
            "MySchemaName": {...}
        },
        "securitySchemes": {
            ...
        },
    },
    "info": {
        "contact": {...},
        "description": ...,
        "title": ...,
        "version": ...,
    },
    "openapi": "3.1.0",
    "paths": {...},
    "servers": [...],
}
```

**Reply by stefanofusai** [reply]
Hey @Julian, did you manage to take a look at this? Thanks!

**Reply by Julian** [reply]
The subschema you showed isn't a valid schema -- I'm not sure what there is to look at on this side, it sounds like you're using a tool to generate the schema, it might have a bug, it's hard to know without the full schema.
